contracts/cw-proposal-multiple/src/contract.rs

```diff
@@ -0,0 +1,715 @@
+    PROPOSAL_COUNT.save(deps.storage, &0)?;
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::default()
+        .add_attribute("action", "instantiate")
+        .add_attribute("dao", config.dao))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response<Empty>, ContractError> {
+    match msg {
+        ExecuteMsg::Propose {
+            title,
+            description,
+            choices,
+        } => execute_propose(deps, env, info.sender, title, description, choices),
+        ExecuteMsg::Vote { proposal_id, vote } => execute_vote(deps, env, info, proposal_id, vote),
+        ExecuteMsg::Execute { proposal_id } => execute_execute(deps, env, info, proposal_id),
+        ExecuteMsg::Close { proposal_id } => execute_close(deps, env, info, proposal_id),
+        ExecuteMsg::UpdateConfig {
+            voting_strategy,
+            min_voting_period,
+            max_voting_period,
+            only_members_execute,
+            dao,
+            deposit_info,
+        } => execute_update_config(
+            deps,
+            info,
+            voting_strategy,
+            min_voting_period,
+            max_voting_period,
+            only_members_execute,
+            dao,
+            deposit_info,
+        ),
+        ExecuteMsg::AddProposalHook { address } => {
+            execute_add_proposal_hook(deps, env, info, address)
+        }
+        ExecuteMsg::RemoveProposalHook { address } => {
+            execute_remove_proposal_hook(deps, env, info, address)
+        }
+        ExecuteMsg::AddVoteHook { address } => execute_add_vote_hook(deps, env, info, address),
+        ExecuteMsg::RemoveVoteHook { address } => {
+            execute_remove_vote_hook(deps, env, info, address)
+        }
+    }
+}
+
+pub fn execute_propose(
+    deps: DepsMut,
+    env: Env,
+    sender: Addr,
+    title: String,
+    description: String,
+    options: MultipleChoiceOptions,
+) -> Result<Response<Empty>, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    let voting_module: Addr = deps
+        .querier
+        .query_wasm_smart(config.dao.clone(), &cw_core::msg::QueryMsg::VotingModule {})?;
+
+    // Voting modules are not required to implement this
+    // query. Lacking an implementation they are active by default.
+    let active_resp: IsActiveResponse = deps
+        .querier
+        .query_wasm_smart(
+            voting_module,
+            &cw_core_interface::voting::Query::IsActive {},
+        )
+        .unwrap_or(IsActiveResponse { active: true });
+
+    if !active_resp.active {
+        return Err(ContractError::InactiveDao {});
+    }
+
+    // Check that the sender is a member of the governance contract.
+    let sender_power = get_voting_power(deps.as_ref(), sender.clone(), config.dao.clone(), None)?;
+    if sender_power.is_zero() {
+        return Err(ContractError::Unauthorized {});
```

Comment:
nit: error could be improved here, could be replaced with a MustHaveVotingPowerToPropose error. verbose but does the job

Reply:
agreed!